File: cs336_basics/tokenizer.py
from __future__ import annotations

import ast
import json
import pickle
from collections.abc import Iterable, Iterator
from pathlib import Path

import regex as re


PAT = r"""'(?:[sdmt]|ll|ve|re)| ?\p{L}+| ?\p{N}+| ?[^\s\p{L}\p{N}]+|\s+(?!\S)|\s+"""
PRETOKEN_RE = re.compile(PAT)
BYTE_TOKENS = tuple(bytes([i]) for i in range(256))

Pair = tuple[bytes, bytes]
TokenSegment = tuple[int, int, str | None]


class Tokenizer:
    @staticmethod
    def _gpt2_byte_decoder() -> dict[str, int]:
        byte_values = list(range(ord("!"), ord("~") + 1))
        byte_values += list(range(ord("¡"), ord("¬") + 1))
        byte_values += list(range(ord("®"), ord("ÿ") + 1))
        code_points = byte_values[:]
        next_shifted = 0
        for byte_value in range(256):
            if byte_value not in byte_values:
                byte_values.append(byte_value)
                code_points.append(256 + next_shifted)
                next_shifted += 1
        return {chr(code_point): byte_value for byte_value, code_point in zip(byte_values, code_points)}

    @staticmethod
    def _decode_gpt2_token(token: str, byte_decoder: dict[str, int]) -> bytes:
        return bytes(byte_decoder[character] for character in token)

    @classmethod
    def _load_vocab(cls, vocab_filepath: str) -> dict[int, bytes]:
        path = Path(vocab_filepath)
        if path.suffix == ".pkl":
            with path.open("rb") as f:
                return pickle.load(f)

        with path.open(encoding="utf-8") as f:
            data = json.load(f)

        if isinstance(data, dict) and isinstance(data.get("tokens"), list):
            vocab: dict[int, bytes] = {}
            for token in data["tokens"]:
                vocab[int(token["id"])] = bytes(token["byte_values"])
            return vocab

        if isinstance(data, dict) and all(isinstance(value, int) for value in data.values()):
            byte_decoder = cls._gpt2_byte_decoder()
            return {token_id: cls._decode_gpt2_token(token, byte_decoder) for token, token_id in data.items()}

        if isinstance(data, dict):
            return {int(token_id): bytes(token_bytes) for token_id, token_bytes in data.items()}

        raise ValueError(f"Unsupported vocabulary file format: {vocab_filepath}")

    @classmethod
    def _load_merges(cls, merges_filepath: str) -> list[Pair]:
        path = Path(merges_filepath)
        if path.suffix == ".pkl":
            with path.open("rb") as f:
                return pickle.load(f)

        byte_decoder = cls._gpt2_byte_decoder()
        merges: list[Pair] = []
        with path.open(encoding="utf-8") as f:
            for line in f:
                line = line.rstrip("\n")
                if not line or line.startswith("#"):
                    continue

                parts = line.split("\t")
                if len(parts) >= 3 and parts[0].isdigit():
                    merges.append((ast.literal_eval(parts[1]), ast.literal_eval(parts[2])))
                    continue

                parts = line.split(" ")
                if len(parts) == 2:
                    merges.append(
                        (
                            cls._decode_gpt2_token(parts[0], byte_decoder),
                            cls._decode_gpt2_token(parts[1], byte_decoder),
                        )
                    )

        return merges

    def __init__(
        self,
        vocab: dict[int, bytes],
        merges: list[Pair],
        special_tokens: list[str] | None = None,
    ) -> None:
        self.vocab = dict(vocab)
        self.token_to_id = {token: token_id for token_id, token in self.vocab.items()}

        self.special_tokens = list(special_tokens) if special_tokens is not None else []
        self.special_token_ids: dict[str, int] = {}
        next_token_id = max(self.vocab, default=-1) + 1
        for special_token in self.special_tokens:
            special_bytes = special_token.encode("utf-8")
            token_id = self.token_to_id.get(special_bytes)
            if token_id is None:
                token_id = next_token_id
                next_token_id += 1
                self.vocab[token_id] = special_bytes
                self.token_to_id[special_bytes] = token_id
            self.special_token_ids[special_token] = token_id

        self.merge_ranks = {pair: rank for rank, pair in enumerate(merges)}
        self.byte_token_ids = tuple(self.token_to_id[token] for token in BYTE_TOKENS)
        self.merge_ranks_by_id: dict[tuple[int, int], int] = {}
        self.merge_output_by_pair_id: dict[tuple[int, int], int] = {}
        for pair, rank in self.merge_ranks.items():
            left, right = pair
            pair_ids = (self.token_to_id[left], self.token_to_id[right])
            self.merge_ranks_by_id[pair_ids] = rank
            self.merge_output_by_pair_id[pair_ids] = self.token_to_id[left + right]
        self._encode_cache: dict[bytes, tuple[int, ...]] = {}
        self._max_cache_size = 50_000

        self._max_special_token_length = max((len(token) for token in self.special_tokens), default=0)
        if self.special_tokens:
            special_pattern = "|".join(re.escape(token) for token in sorted(self.special_tokens, key=len, reverse=True))
            self._special_re: re.Pattern[str] | None = re.compile(special_pattern)
        else:
            self._special_re = None

    @classmethod
    def from_files(
        cls,
        vocab_filepath: str,
        merges_filepath: str,
        special_tokens: list[str] | None = None,
    ) -> Tokenizer:
        vocab = cls._load_vocab(vocab_filepath)
        merges = cls._load_merges(merges_filepath)
        return cls(vocab, merges, special_tokens)

    def encode(self, text: str) -> list[int]:
        return list(self._encode_text(text))

    def encode_iterable(self, iterable: Iterable[str]) -> Iterator[int]:
        buffer = ""
        for chunk in iterable:
            if not chunk:
                continue
            buffer += chunk
            segments = list(self._token_segments(buffer))
            flush_index = self._stream_flush_index_from_segments(buffer, segments)
            if flush_index > 0:
                yield from self._encode_prefix_with_context(buffer, flush_index, segments)
                buffer = buffer[flush_index:]
        if buffer:
            yield from self._encode_text(buffer)

    def decode(self, ids: list[int]) -> str:
        return b"".join(self.vocab[token_id] for token_id in ids).decode("utf-8", errors="replace")

    def _encode_text(self, text: str) -> Iterator[int]:
        if self._special_re is None:
            yield from self._encode_normal_text(text)
            return

        start = 0
        for match in self._special_re.finditer(text):
            if match.start() > start:
                yield from self._encode_normal_text(text[start : match.start()])
            yield self.special_token_ids[match.group(0)]
            start = match.end()
        if start < len(text):
            yield from self._encode_normal_text(text[start:])

    def _encode_normal_text(self, text: str) -> Iterator[int]:
        for match in PRETOKEN_RE.finditer(text):
            yield from self._encode_pretoken(match.group(0).encode("utf-8"))

    def _encode_prefix_with_context(
        self,
        text: str,
        end_index: int,
        segments: Iterable[TokenSegment] | None = None,
    ) -> Iterator[int]:
        if segments is None:
            segments = self._token_segments(text)

        for start, end, special_token in segments:
            if end <= end_index:
                if special_token is None:
                    yield from self._encode_pretoken(text[start:end].encode("utf-8"))
                else:
                    yield self.special_token_ids[special_token]
            elif start < end_index:
                raise ValueError("stream flush boundary split a token")
            else:
                break

    def _encode_pretoken(self, pretoken: bytes) -> tuple[int, ...]:
        cached = self._encode_cache.get(pretoken)
        if cached is not None:
            return cached

        tokens = [self.byte_token_ids[byte] for byte in pretoken]
        while len(tokens) > 1:
            best_pair: tuple[int, int] | None = None
            best_rank: int | None = None
            for i in range(len(tokens) - 1):
                pair = (tokens[i], tokens[i + 1])
                rank = self.merge_ranks_by_id.get(pair)
                if rank is not None and (best_rank is None or rank < best_rank):
                    best_pair = pair
                    best_rank = rank

            if best_pair is None:
                break

            merged_token = self.merge_output_by_pair_id[best_pair]
            merged_tokens: list[int] = []
            i = 0
            while i < len(tokens):
                if i + 1 < len(tokens) and tokens[i] == best_pair[0] and tokens[i + 1] == best_pair[1]:
                    merged_tokens.append(merged_token)
                    i += 2
                else:
                    merged_tokens.append(tokens[i])
                    i += 1
            tokens = merged_tokens

        token_ids = tuple(tokens)
        if len(self._encode_cache) >= self._max_cache_size:
            self._encode_cache.clear()
        self._encode_cache[pretoken] = token_ids
        return token_ids

    def _stream_flush_index(self, text: str) -> int:
        return self._stream_flush_index_from_segments(text, list(self._token_segments(text)))

    def _stream_flush_index_from_segments(self, text: str, segments: list[TokenSegment]) -> int:
        if not segments:
            return 0

        keep_start = segments[-1][0]
        if self._max_special_token_length > 1:
            keep_start = min(keep_start, max(0, len(text) - self._max_special_token_length + 1))

        for start, end, _ in segments:
            if start < keep_start < end:
                return start
        return keep_start

    def _token_spans(self, text: str) -> Iterator[tuple[int, int]]:
        for start, end, _ in self._token_segments(text):
            yield start, end

    def _token_segments(self, text: str) -> Iterator[TokenSegment]:
        if self._special_re is None:
            for match in PRETOKEN_RE.finditer(text):
                yield match.start(), match.end(), None
            return

        start = 0
        for special_match in self._special_re.finditer(text):
            if special_match.start() > start:
                for match in PRETOKEN_RE.finditer(text, start, special_match.start()):
                    yield match.start(), match.end(), None
            yield special_match.start(), special_match.end(), special_match.group(0)
            start = special_match.end()
        if start < len(text):
            for match in PRETOKEN_RE.finditer(text, start):
                yield match.start(), match.end(), None

File: crates/cs336_bpe_rs/src/encoder/tokenizer.rs
use std::collections::HashMap;
use std::path::Path;

use anyhow::{anyhow, bail, Context, Error, Result};

use crate::encoder::merges::load_merges;
use crate::encoder::streaming::TokenSegment;
use crate::encoder::vocab::load_vocab;
use crate::pretokenizer::{
    find_next_special, pretoken_byte_strings, pretoken_spans, sorted_special_tokens,
};
use crate::trainer::state::{BytePair, TokenId, TokenPair};

const MAX_CACHE_SIZE: usize = 50_000;

#[derive(Debug, Clone)]
pub struct Tokenizer {
    vocab: Vec<Vec<u8>>,
    token_to_id: HashMap<Vec<u8>, TokenId>,
    special_tokens: Vec<String>,
    sorted_special_tokens: Vec<String>,
    special_token_ids: HashMap<String, TokenId>,
    merge_ranks_by_id: HashMap<TokenPair, usize>,
    merge_output_by_pair_id: HashMap<TokenPair, TokenId>,
    byte_token_ids: [TokenId; 256],
    encode_cache: HashMap<Vec<u8>, Vec<TokenId>>,
    max_special_token_length: usize,
}

impl Tokenizer {
    pub fn from_files(
        vocab_filepath: impl AsRef<Path>,
        merges_filepath: impl AsRef<Path>,
        special_tokens: Vec<String>,
    ) -> Result<Self> {
        let vocab = load_vocab(vocab_filepath.as_ref())?;
        let merges = load_merges(merges_filepath.as_ref())?;
        Self::new(vocab, merges, special_tokens)
    }

    pub fn new(
        mut vocab: Vec<Vec<u8>>,
        merges: Vec<BytePair>,
        special_tokens: Vec<String>,
    ) -> Result<Self> {
        let mut token_to_id = HashMap::new();
        for (token_id, token) in vocab.iter().enumerate() {
            token_to_id.insert(token.clone(), token_id as TokenId);
        }

        let mut special_token_ids = HashMap::new();
        for special_token in &special_tokens {
            let special_bytes = special_token.as_bytes().to_vec();
            let token_id = match token_to_id.get(&special_bytes).copied() {
                Some(token_id) => token_id,
                None => {
                    let token_id = vocab.len() as TokenId;
                    vocab.push(special_bytes.clone());
                    token_to_id.insert(special_bytes, token_id);
                    token_id
                }
            };
            special_token_ids.insert(special_token.clone(), token_id);
        }

        let mut byte_token_ids = [0; 256];
        for byte in 0u16..=255 {
            let token = vec![byte as u8];
            byte_token_ids[byte as usize] = *token_to_id
                .get(&token)
                .with_context(|| format!("vocabulary missing byte token {byte}"))?;
        }

        let mut merge_ranks_by_id = HashMap::new();
        let mut merge_output_by_pair_id = HashMap::new();
        for (rank, (left, right)) in merges.iter().enumerate() {
            let left_id = *token_to_id
                .get(left)
                .with_context(|| format!("merge left token missing from vocab: {left:?}"))?;
            let right_id = *token_to_id
                .get(right)
                .with_context(|| format!("merge right token missing from vocab: {right:?}"))?;
            let mut merged = left.clone();
            merged.extend(right);
            let merged_id = *token_to_id
                .get(&merged)
                .with_context(|| format!("merge output token missing from vocab: {merged:?}"))?;
            let pair = (left_id, right_id);
            merge_ranks_by_id.insert(pair, rank);
            merge_output_by_pair_id.insert(pair, merged_id);
        }

        let sorted_special_tokens = sorted_special_tokens(&special_tokens);
        let max_special_token_length = special_tokens
            .iter()
            .map(|token| token.len())
            .max()
            .unwrap_or(0);

        Ok(Self {
            vocab,
            token_to_id,
            special_tokens,
            sorted_special_tokens,
            special_token_ids,
            merge_ranks_by_id,
            merge_output_by_pair_id,
            byte_token_ids,
            encode_cache: HashMap::new(),
            max_special_token_length,
        })
    }

    pub fn encode(&mut self, text: &str) -> Result<Vec<TokenId>> {
        self.encode_text(text)
    }

    pub fn encode_iterable<I>(&mut self, iterable: I) -> Result<Vec<TokenId>>
    where
        I: IntoIterator<Item = String>,
    {
        let iterable = iterable.into_iter().map(Ok::<String, Error>);
        let mut output = Vec::new();
        self.encode_iterable_result_to_sink(iterable, |token_id| {
            output.push(token_id);
            Ok(())
        })?;
        Ok(output)
    }

    pub fn encode_iterable_result_to_sink<I, E, F>(
        &mut self,
        iterable: I,
        mut sink: F,
    ) -> Result<()>
    where
        I: IntoIterator<Item = std::result::Result<String, E>>,
        E: std::fmt::Display,
        F: FnMut(TokenId) -> Result<()>,
    {
        let mut buffer = String::new();
        for chunk in iterable {
            let chunk = chunk.map_err(|error| anyhow!("{error}"))?;
            if chunk.is_empty() {
                continue;
            }
            buffer.push_str(&chunk);
            let segments = self.token_segments(&buffer)?;
            let flush_index = self.stream_flush_index_from_segments(&buffer, &segments);
            if flush_index > 0 {
                for token_id in self.encode_prefix_with_context(&buffer, flush_index, &segments)? {
                    sink(token_id)?;
                }
                buffer = buffer[flush_index..].to_string();
            }
        }
        if !buffer.is_empty() {
            for token_id in self.encode_text(&buffer)? {
                sink(token_id)?;
            }
        }
        Ok(())
    }

    pub fn decode(&self, ids: &[TokenId]) -> Result<String> {
        let mut bytes = Vec::new();
        for &token_id in ids {
            let token = self
                .vocab
                .get(token_id as usize)
                .with_context(|| format!("unknown token id {token_id}"))?;
            bytes.extend(token);
        }
        Ok(String::from_utf8_lossy(&bytes).into_owned())
    }

    fn encode_text(&mut self, text: &str) -> Result<Vec<TokenId>> {
        let mut ids = Vec::new();
        if self.sorted_special_tokens.is_empty() {
            ids.extend(self.encode_normal_text(text)?);
            return Ok(ids);
        }

        let mut start = 0;
        while let Some((special_start, special_end, special_token)) =
            find_next_special(text, start, &self.sorted_special_tokens)
        {
            if special_start > start {
                ids.extend(self.encode_normal_text(&text[start..special_start])?);
            }
            ids.push(
                *self
                    .special_token_ids
                    .get(&special_token)
                    .with_context(|| format!("missing special token id for {special_token}"))?,
            );
            start = special_end;
        }
        if start < text.len() {
            ids.extend(self.encode_normal_text(&text[start..])?);
        }
        Ok(ids)
    }

    fn encode_normal_text(&mut self, text: &str) -> Result<Vec<TokenId>> {
        let mut ids = Vec::new();
        for pretoken in pretoken_byte_strings(text)? {
            ids.extend(self.encode_pretoken(&pretoken)?);
        }
        Ok(ids)
    }

    pub fn encode_pretoken(&mut self, pretoken: &[u8]) -> Result<Vec<TokenId>> {
        if let Some(cached) = self.encode_cache.get(pretoken) {
            return Ok(cached.clone());
        }

        let mut tokens: Vec<TokenId> = pretoken
            .iter()
            .map(|byte| self.byte_token_ids[*byte as usize])
            .collect();

        while tokens.len() > 1 {
            let mut best_pair: Option<TokenPair> = None;
            let mut best_rank: Option<usize> = None;
            for window in tokens.windows(2) {
                let pair = (window[0], window[1]);
                if let Some(&rank) = self.merge_ranks_by_id.get(&pair) {
                    if best_rank.is_none_or(|current| rank < current) {
                        best_pair = Some(pair);
                        best_rank = Some(rank);
                    }
                }
            }

            let Some(best_pair) = best_pair else {
                break;
            };
            let merged_token = *self
                .merge_output_by_pair_id
                .get(&best_pair)
                .context("merge rank exists without merge output")?;
            let mut merged_tokens = Vec::with_capacity(tokens.len());
            let mut i = 0;
            while i < tokens.len() {
                if i + 1 < tokens.len() && tokens[i] == best_pair.0 && tokens[i + 1] == best_pair.1
                {
                    merged_tokens.push(merged_token);
                    i += 2;
                } else {
                    merged_tokens.push(tokens[i]);
                    i += 1;
                }
            }
            tokens = merged_tokens;
        }

        if self.encode_cache.len() >= MAX_CACHE_SIZE {
            self.encode_cache.clear();
        }
        self.encode_cache.insert(pretoken.to_vec(), tokens.clone());
        Ok(tokens)
    }

    fn encode_prefix_with_context(
        &mut self,
        text: &str,
        end_index: usize,
        segments: &[TokenSegment],
    ) -> Result<Vec<TokenId>> {
        let mut ids = Vec::new();
        for (start, end, special_token) in segments {
            if *end <= end_index {
                if let Some(special_token) = special_token {
                    ids.push(*self.special_token_ids.get(special_token).with_context(|| {
                        format!("missing special token id for {special_token}")
                    })?);
                } else {
                    ids.extend(self.encode_pretoken(text[*start..*end].as_bytes())?);
                }
            } else if *start < end_index {
                bail!("stream flush boundary split a token");
            } else {
                break;
            }
        }
        Ok(ids)
    }

    fn stream_flush_index_from_segments(&self, text: &str, segments: &[TokenSegment]) -> usize {
        let Some((last_start, _, _)) = segments.last() else {
            return 0;
        };

        let mut keep_start = *last_start;
        if self.max_special_token_length > 1 {
            let special_keep_start = text
                .len()
                .saturating_sub(self.max_special_token_length)
                .saturating_add(1);
            keep_start = keep_start.min(previous_char_boundary(text, special_keep_start));
        }

        for (start, end, _) in segments {
            if *start < keep_start && keep_start < *end {
                return *start;
            }
        }
        keep_start
    }

    fn token_segments(&self, text: &str) -> Result<Vec<TokenSegment>> {
        let mut segments = Vec::new();
        if self.sorted_special_tokens.is_empty() {
            for (start, end) in pretoken_spans(text)? {
                segments.push((start, end, None));
            }
            return Ok(segments);
        }

        let mut start = 0;
        while let Some((special_start, special_end, special_token)) =
            find_next_special(text, start, &self.sorted_special_tokens)
        {
            if special_start > start {
                for (relative_start, relative_end) in pretoken_spans(&text[start..special_start])? {
                    segments.push((start + relative_start, start + relative_end, None));
                }
            }
            segments.push((special_start, special_end, Some(special_token)));
            start = special_end;
        }
        if start < text.len() {
            for (relative_start, relative_end) in pretoken_spans(&text[start..])? {
                segments.push((start + relative_start, start + relative_end, None));
            }
        }
        Ok(segments)
    }

    pub fn vocab(&self) -> &[Vec<u8>] {
        &self.vocab
    }

    pub fn special_tokens(&self) -> &[String] {
        &self.special_tokens
    }

    pub fn token_to_id_len(&self) -> usize {
        self.token_to_id.len()
    }
}

fn previous_char_boundary(text: &str, mut index: usize) -> usize {
    index = index.min(text.len());
    while index > 0 && !text.is_char_boundary(index) {
        index -= 1;
    }
    index
}

#[cfg(test)]
mod tests {
    use super::Tokenizer;

    fn tiny_tokenizer() -> Tokenizer {
        let mut vocab: Vec<Vec<u8>> = (0u16..=255).map(|value| vec![value as u8]).collect();
        vocab.push(b"ab".to_vec());
        vocab.push(b"abc".to_vec());
        Tokenizer::new(
            vocab,
            vec![
                (b"a".to_vec(), b"b".to_vec()),
                (b"ab".to_vec(), b"c".to_vec()),
            ],
            vec!["<|endoftext|>".to_string()],
        )
        .unwrap()
    }

    #[test]
    fn encodes_pretoken_by_lowest_rank_merge() {
        let mut tokenizer = tiny_tokenizer();
        assert_eq!(tokenizer.encode_pretoken(b"abc").unwrap(), vec![257]);
    }

    #[test]
    fn preserves_special_tokens_with_longest_match() {
        let mut vocab: Vec<Vec<u8>> = (0u16..=255).map(|value| vec![value as u8]).collect();
        vocab.push(b"<|endoftext|>".to_vec());
        vocab.push(b"<|endoftext|><|endoftext|>".to_vec());
        let mut tokenizer = Tokenizer::new(
            vocab,
            Vec::new(),
            vec![
                "<|endoftext|>".to_string(),
                "<|endoftext|><|endoftext|>".to_string(),
            ],
        )
        .unwrap();
        let ids = tokenizer
            .encode("a<|endoftext|><|endoftext|>b<|endoftext|>")
            .unwrap();
        assert!(ids.contains(&257));
        assert!(ids.contains(&256));
    }

    #[test]
    fn decodes_with_utf8_replacement() {
        let tokenizer = tiny_tokenizer();
        assert_eq!(tokenizer.decode(&[255]).unwrap(), "\u{fffd}");
    }

    #[test]
    fn streaming_matches_whole_input() {
        let mut tokenizer = tiny_tokenizer();
        let text = "abc abc<|endoftext|>abc";
        let whole = tokenizer.encode(text).unwrap();
        for chunk_size in [1, 2, 7] {
            let chunks = text
                .as_bytes()
                .chunks(chunk_size)
                .map(|chunk| String::from_utf8(chunk.to_vec()).unwrap())
                .collect::<Vec<_>>();
            let mut streaming_tokenizer = tiny_tokenizer();
            assert_eq!(streaming_tokenizer.encode_iterable(chunks).unwrap(), whole);
        }
    }
}
